### ISSUE TITLE
Reuse the logits in the _prepare_inputs.

### ISSUE THREAD

**Issue by linkedlist771** [huggingface/trl#2888]
### Feature request

In the `_prepare_inputs`, the [code](https://github.com/huggingface/trl/blob/main/trl/trainer/grpo_trainer.py#L542) have already calculated the completion tokens(using vllm or the HF generate), which carries the logits information. 
```python
completion_ids = [out.token_ids for completions in outputs for out in completions.outputs]
```

Why bother to calculate it again in the [compute_loss](https://github.com/huggingface/trl/blob/main/trl/trainer/grpo_trainer.py#L700)
```python
per_token_logps = self._get_per_token_logps(model, input_ids, attention_mask, logits_to_keep)
```

### Motivation

reuse the logits can speed the training.

### Your contribution

I may try to submit a PR, no idea yet.

**Comment by albertvillanova** [comment]
Hey @linkedlist771, sorry for the very slow reply here — we're catching up on the backlog. 

Good question. The recompute in compute_loss is intentional: the per_token_logps used in the policy loss must be differentiable w.r.t. the current model, but the logprobs produced during generation (vLLM or HF generate) are detached and come from a separate numerical path, so they can't carry gradients for backprop and can't replace the loss-time term (see https://github.com/huggingface/trl/blob/d71a6ca90806bd6dc62cdea5cc26b69f4c376403/trl/trainer/grpo_trainer.py#L2637). 

That said, your underlying idea — reuse the generation logprobs — is exactly what TRL does now: we capture the sampling-time logprobs and use them for vLLM importance-sampling correction to fix the train/inference distribution mismatch (https://github.com/huggingface/trl/blob/d71a6ca90806bd6dc62cdea5cc26b69f4c376403/trl/trainer/grpo_trainer.py#L2204-L2215, enable with vllm_importance_sampling_correction). So the cheap, useful part of the reuse is already shipped, while the gradient-bearing forward still has to run. 

Closing as working-as-intended, but happy to reopen if you had a specific speedup in mind that this doesn't cover.
